concordium-smart-contract-testing/src/types.rs

```diff
@@ -360,6 +378,55 @@ impl ContractInvokeSuccess {
     }
 }
 
+/// A wrapper for [`ContractTraceElement`], which provides additional
+/// information for testing and debugging. Most notably, it contains trace
+/// elements for failures, which are normally discarded by the node.
+#[derive(Debug)]
+pub enum DebugTraceElement {
+    /// A regular trace element with some additional data, e.g., energy usage
+    /// and the entrypoint.
+    /// This variant may be included included in the `WithFailures` list of
+    /// trace elements.
+    Regular {
+        /// The entrypoint.
+        entrypoint:    OwnedEntrypointName,
+        /// The trace element.
+        trace_element: ContractTraceElement,
+        /// The energy used so far.
+        energy_used:   Energy,
+    },
+    /// One or multiple trace elements that fail. Useful for debugging.
+    /// This variant also contains additional information, such as the error,
+    /// entrypoint, and energy usage.
+    WithFailures {
+        /// The address of the contract which failed.
+        contract_address: ContractAddress,
+        /// The entrypoint which failed.
+        entrypoint:       OwnedEntrypointName,
+        /// The error returned.
+        error:            InvokeExecutionError,
+        /// Intermediate [`DebugTraceElement`]s which occured prior to failing.
+        /// These are the elements which are normally discared by the node.
+        trace_elements:   Vec<DebugTraceElement>,
+        /// The energy used so far.
+        energy_used:      Energy,
+    },
+}
+
+/// The reason for why a contract invocation failed during execution.
+#[derive(Debug)]
+pub enum InvokeExecutionError {
+    /// The contract rejected.
+    Reject {
+        /// The error code for why it rejected.
+        reason:       i32,
+        /// The return value.
+        return_value: ReturnValue,
+    },
+    /// The contract trapped.
+    Trap { error: ExecutionError },
```

Comment:
Just for my understanding:
`Trap` happens if e.g. the transaction runs out of `energy` or if the logic of the smart contract reverts without a specific reject reason `reject error code`, isn't it?

Reply:
Almost correct. Out of energy is a separate kind of error, which is not included here since it will be caught differently and has its own error kind in `InvokeErrorKind`.
But yes, the traps occur when something unexpected happens in the contract, fx an `unwrap` that panics or similar.